src/kokkos_abstraction.hpp

```diff
@@ -258,6 +258,35 @@ par_dispatch(LoopPatternMDRange, const std::string &name, DevExecSpace exec_spac
                   function, std::forward<Args>(args)...);
 }
 
+template <typename, typename>
+class FlatFunctor;
+
+template <typename F, typename... Args>
+auto MakeFlatFunctor(F &function, Args... args) {
+  return FlatFunctor<F, decltype(&F::operator())>(function, std::forward<Args>(args)...);
+}
+
+template <typename Function, typename R, typename T, typename Index, typename... FArgs>
+class FlatFunctor<Function, R (T::*)(Index, Index, Index, FArgs...) const> {
+  int NjNi, Nj, Ni, kl, jl, il;
+  Function function;
+
+ public:
+  FlatFunctor(const Function _function, const int _NjNi, const int _Nj, const int _Ni,
+              const int _kl, const int _jl, const int _il)
+      : function(_function), NjNi(_NjNi), Nj(_Nj), Ni(_Ni), kl(_kl), jl(_jl), il(_il) {}
+  KOKKOS_INLINE_FUNCTION
+  void operator()(const int &idx, FArgs &&...fargs) const {
+    int k = idx / NjNi;
+    int j = (idx - k * NjNi) / Ni;
+    int i = idx - k * NjNi - j * Ni;
+    k += kl;
+    j += jl;
+    i += il;
+    function(k, j, i, std::forward<FArgs>(fargs)...);
+  }
+};
```

Comment:
Probably not for this PR, but I wonder if we could make this work for an arbitrary dimensional index space using `IndexRange` in `utils/` with a little bit of template magic.

Reply:
I think it could work. This seems to work and at least gets part way there and seems to do what I would expect https://godbolt.org/z/fbYP5vW6r

Reply:
This (https://godbolt.org/z/n48951jGK) seems to work, but I didn't explicitly pull out the function signature like you did (which I think matters for references).

Reply:
yeah I remember cuda builds had given me trouble with that. This [one](https://godbolt.org/z/4W37ejM7E) uses some more template helper structs to also handle `IndexRange`s in the constructor. I imagine you could use similar things to handle all the `par_dispatch` overloads as well

Reply:
Saved for future reference here https://github.com/parthenon-hpc-lab/parthenon/issues/1134